libraries/ESP8266WiFi/src/CertStoreSPIFFSBearSSL.cpp

```diff
@@ -0,0 +1,222 @@
+  int sz = f.size();
+  uint8_t *buf = new uint8_t[sz];
+  if (!buf) {
+    f.close();
+    return ci;
+  }
+  f.read(buf, sz);
+  f.close();
+
+  // Process it using SHA256, same as the hashed_dn
+  br_x509_decoder_context *ctx = new br_x509_decoder_context;
+  br_sha256_context *sha256 = new br_sha256_context;
+  br_sha256_init(sha256);
+  br_x509_decoder_init(ctx, dn_append, sha256, nullptr, nullptr);
+  br_x509_decoder_push(ctx, (const void*)buf, sz);
+
+  // Copy result to structure
+  br_sha256_out(sha256, &ci.sha256);
+  strcpy(ci.fname, fname);
+
+  delete sha256;
+  delete ctx;
+  delete buf;
+  return ci;
+}
+
+void CertStoreSPIFFSBearSSL::initCertStore()
+{
+  SPIFFS.begin();
+  File tbl = SPIFFS.open("/ca_tbl.bin", "w");
+  if (!tbl) return;
+  Dir d = SPIFFS.openDir("");
+  while (d.next()) {
+    if (!strstr(d.fileName().c_str(), ".der")) continue;
+    CertStoreSPIFFSBearSSL::CertInfo ci = preprocessCert(d.fileName().c_str());
+    tbl.write((uint8_t*)&ci, sizeof(ci));
+  }
+  tbl.close();
+}
+
+void CertStoreSPIFFSBearSSL::installCertStore(br_x509_minimal_context *ctx)
+{
+  br_x509_minimal_set_dynamic(ctx, findHashedTA, freeHashedTA);
+}
+
+
+const br_x509_trust_anchor *CertStoreSPIFFSBearSSL::findHashedTA(void *hashed_dn, size_t len)
+{
+  CertInfo ci;
+
+  if (len != sizeof(ci.sha256) || !SPIFFS.exists("/ca_tbl.bin")) {
+    return nullptr;
+  }
+
+  File f = SPIFFS.open("/ca_tbl.bin", "r");
+  if (!f) {
+    return nullptr;
+  }
+  while (f.read((uint8_t*)&ci, sizeof(ci)) == sizeof(ci)) {
+    if (!memcmp(ci.sha256, hashed_dn, sizeof(ci.sha256))) {
+      // This could be the one!
+      f.close();
+      File d = SPIFFS.open(ci.fname, "r");
+      if (!d) {
+        return nullptr;
+      }
+      size_t der_len = d.size();
+      uint8_t *der = (uint8_t*)malloc(der_len);
+      if (!der) {
+        d.close();
+        return nullptr;
+      }
+      if (d.read(der, der_len) != der_len) {
+        d.close();
+        free(der);
+        return nullptr;
+      }
+      d.close();
+
+      br_x509_decoder_context *dc = new br_x509_decoder_context;
+      br_x509_decoder_init(dc, dn_append_null, nullptr, nullptr, nullptr);
+      br_x509_decoder_push(dc, der, der_len);
+      free(der);
+      br_x509_pkey *pk = br_x509_decoder_get_pkey(dc);
+      if (!pk) {
+        delete dc;
+        return nullptr;
+      }
+
+      br_x509_trust_anchor *ta = (br_x509_trust_anchor*)malloc(sizeof(br_x509_trust_anchor));
+      if (!ta) {
+        delete dc;
+        return nullptr;
+      }
+      memset(ta, 0, sizeof(*ta));
+      ta->dn.data = (uint8_t*)malloc(sizeof(ci.sha256));
+      if (!ta->dn.data) {
+        free(ta);
+        delete dc;
+        return nullptr;
+      }
+      memcpy(ta->dn.data, ci.sha256, sizeof(ci.sha256));
+      ta->dn.len = sizeof(ci.sha256);
+      ta->flags = 0;
+      if (br_x509_decoder_isCA(dc)) {
+        ta->flags |= BR_X509_TA_CA;
+      }
+
+      delete dc; // Done with x509 decoder
```

Comment:
At this point pk still holds reference to structure inside dc, so it cannot be freed here.
Issue became visible when add some debug prints after this line corrupted the private key.

Reply:
Thanks for finding this!  I'll throw a new commit up soon, simple fix (you've probably already done it in your own copy).